kubernetes_cluster_node_pool: Fix race condition with virtual network status when creating node pool

c4milo opened the pull request:
Related to https://github.com/hashicorp/terraform-provider-azurerm/issues/13105. We are able to consistently trigger it in our testing. 

```
❯  make acctests SERVICE='containers' TESTARGS='-run=TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/containers -run=TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition
=== PAUSE TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition
=== CONT  TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition
--- PASS: TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition (1395.32s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/containers	1398.094s
```

### Error 
```console
    Error: creating Agent Pool (Subscription: "312d56f6-697b-4377-9a9a-83257ce33066"
        Resource Group Name: "acctestRG-aks-240506121623880995"
        Managed Cluster Name: "acctestaks240506121623880995"
        Agent Pool Name: "internal3"): polling after CreateOrUpdate: polling failed: the Azure API returned the following error:

        Status: "SetVNetOwnershipFailed"
        Code: ""
        Message: "Set virtual network ownership failed. Subscription: 312d56f6-697b-4377-9a9a-83257ce33066; resource group: acctestRG-aks-240506121623880995; virtual network name: acctestvirtnet240506121623880995. autorest/azure: Service returned an error. Status=400 Code=\"VirtualNetworkNotInSucceededState\" Message=\"Virtual network /subscriptions/312d56f6-697b-4377-9a9a-83257ce33066/resourceGroups/acctestRG-aks-240506121623880995/providers/Microsoft.Network/virtualNetworks/acctestvirtnet240506121623880995 is in Updating state. It needs to be in Succeeded state in order to set resource ownership.\" Details=[]\nVirtual network /subscriptions/312d56f6-697b-4377-9a9a-83257ce33066/resourceGroups/acctestRG-aks-240506121623880995/providers/Microsoft.Network/virtualNetworks/acctestvirtnet240506121623880995 is in Updating state. It needs to be in Succeeded state in order to set resource ownership."
        Activity Id: ""

        ---

        API Response:

        ----[start]----
        {
          "name": "da6f9f40-6f18-4380-b252-d516cd3659f0",
          "status": "Failed",
          "startTime": "2024-05-06T16:21:01.6132767Z",
          "endTime": "2024-05-06T16:21:09.0898922Z",
          "error": {
           "code": "SetVNetOwnershipFailed",
           "message": "Set virtual network ownership failed. Subscription: 312d56f6-697b-4377-9a9a-83257ce33066; resource group: acctestRG-aks-240506121623880995; virtual network name: acctestvirtnet240506121623880995. autorest/azure: Service returned an error. Status=400 Code=\"VirtualNetworkNotInSucceededState\" Message=\"Virtual network /subscriptions/312d56f6-697b-4377-9a9a-83257ce33066/resourceGroups/acctestRG-aks-240506121623880995/providers/Microsoft.Network/virtualNetworks/acctestvirtnet240506121623880995 is in Updating state. It needs to be in Succeeded state in order to set resource ownership.\" Details=[]",
           "details": [
            {
             "code": "",
             "message": "Virtual network /subscriptions/312d56f6-697b-4377-9a9a-83257ce33066/resourceGroups/acctestRG-aks-240506121623880995/providers/Microsoft.Network/virtualNetworks/acctestvirtnet240506121623880995 is in Updating state. It needs to be in Succeeded state in order to set resource ownership."
            }
           ]
          }
         }
        -----[end]-----


          with azurerm_kubernetes_cluster_node_pool.test3,
          on terraform_plugin_test.tf line 97, in resource "azurerm_kubernetes_cluster_node_pool" "test3":
          97: resource "azurerm_kubernetes_cluster_node_pool" "test3" {
```

<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->


## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.



## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->


## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_kubernetes_cluster_node_pool` - Fix race condition between virtual network status and node pool creation. The virtual network must be in Succeeded state for an AKS node pool to be correctly created. 


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [x] Bug Fix


## Related Issue(s)
https://github.com/hashicorp/terraform-provider-azurerm/issues/13105


> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

stephybun commented on internal/services/containers/kubernetes_cluster_node_pool_resource.go:
```diff
@@ -435,6 +442,17 @@ func resourceKubernetesClusterNodePoolCreate(d *pluginsdk.ResourceData, meta int
 
+	subnetID, err := commonids.ParseSubnetID(d.Get("vnet_subnet_id").(string))
+	if err != nil {
+		return fmt.Errorf("parsing subnet ID %+v", err)
```
The parser functions return plenty information in the error message, so there is no need to wrap them
```suggestion
		return err
```

stephybun commented on internal/services/containers/kubernetes_cluster_node_pool_resource.go:
```diff
@@ -26,14 +26,17 @@ import (
 	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/timeouts"
 	"github.com/hashicorp/terraform-provider-azurerm/utils"
+	azureNetwork "github.com/tombuildsstuff/kermit/sdk/network/2022-07-01/network"
```
We're migrating away from Azure's Track 1 SDK. We can import 

```suggestion
	"github.com/hashicorp/go-azure-sdk/resource-manager/network/2023-09-01/subnets"
```
and use the provisioning state enums from there instead.

stephybun commented on internal/services/containers/kubernetes_cluster_node_pool_resource.go:
```diff
@@ -677,6 +695,32 @@ func resourceKubernetesClusterNodePoolCreate(d *pluginsdk.ResourceData, meta int
+	timeout, _ := ctx.Deadline()
+
+	stateConf := &pluginsdk.StateChangeConf{
+		Pending:    []string{string(azureNetwork.ProvisioningStateUpdating)},
+		Target:     []string{string(azureNetwork.ProvisioningStateSucceeded)},
```
```suggestion
		Pending:    []string{string(subnets.ProvisioningStateUpdating)},
		Target:     []string{string(subnets.ProvisioningStateSucceeded)},
```

stephybun commented on internal/services/containers/kubernetes_cluster_node_pool_resource.go:
```diff
@@ -677,6 +695,32 @@ func resourceKubernetesClusterNodePoolCreate(d *pluginsdk.ResourceData, meta int
+
+	vnetId := commonids.NewVirtualNetworkID(subnetID.SubscriptionId, subnetID.ResourceGroupName, subnetID.VirtualNetworkName)
+	vnetStateConf := &pluginsdk.StateChangeConf{
+		Pending:    []string{string(azureNetwork.ProvisioningStateUpdating)},
+		Target:     []string{string(azureNetwork.ProvisioningStateSucceeded)},
```
```suggestion
		Pending:    []string{string(subnets.ProvisioningStateUpdating)},
		Target:     []string{string(subnets.ProvisioningStateSucceeded)},
```

stephybun commented on internal/services/containers/kubernetes_cluster_node_pool_resource.go:
```diff
@@ -435,6 +442,17 @@ func resourceKubernetesClusterNodePoolCreate(d *pluginsdk.ResourceData, meta int
 		return err
 	}
 
+	subnetID, err := commonids.ParseSubnetID(d.Get("vnet_subnet_id").(string))
```
SubnetID isn't always going to be specified so we should probably wrap all of this logic in a check 
```suggestion
	if subnetId, ok := d.GetOk("vnet_subnet_id").(string); ok && subnetId != "" {
	...
	}
```

c4milo replied:
gotcha, makes sense!

stephybun reviewed (changes requested):
Thanks for this PR @c4milo.

I left some review comments and suggestions in-line. Could you take a look and fix those up? Once that's done we can run the rests and give this another review.

c4milo commented:
@stephybun, I've made the changes you kindly requested. Thanks for reviewing!

tombuildsstuff commented on internal/services/containers/kubernetes_cluster_node_pool_resource_test.go:
```diff
@@ -2945,3 +2959,100 @@ resource "azurerm_kubernetes_cluster_node_pool" "test" {
+}
+
+resource "azurerm_subnet" "test" {
+  count                = 20
```
Looks like we could use 8 rather than 20 here?

```suggestion
  count                = 8
```

c4milo replied:
Sure, I used 20 because it was the number of subnets that made the problem visible for us.

tombuildsstuff commented on internal/services/containers/kubernetes_cluster_node_pool_resource.go:
```diff
@@ -677,6 +698,34 @@ func resourceKubernetesClusterNodePoolCreate(d *pluginsdk.ResourceData, meta int
 
+	if subnetID != nil {
+		// Wait for vnet to come back to Succeeded before releasing any locks
+		timeout, _ := ctx.Deadline()
```
can we add a sanity check to this:

```suggestion
		timeout, ok := ctx.Deadline()
		if !ok {
			return fmt.Errorf("internal-error: context had no deadline")
		}
```

tombuildsstuff commented on internal/services/containers/kubernetes_cluster_node_pool_resource.go:
```diff
@@ -677,6 +698,34 @@ func resourceKubernetesClusterNodePoolCreate(d *pluginsdk.ResourceData, meta int
+		// Wait for vnet to come back to Succeeded before releasing any locks
+		timeout, _ := ctx.Deadline()
+
+		stateConf := &pluginsdk.StateChangeConf{
```
This logic exists in a few places, whilst this isn't a blocker for this instance - can we add a TODO to refactor this out into a separate `custompoller` so that we can reuse this in the future:

```suggestion
		// TODO: refactor this into a `custompoller` within the `network` package
		stateConf := &pluginsdk.StateChangeConf{
```

tombuildsstuff reviewed (changes requested):
hey @c4milo 

Thanks for pushing those changes, I've taken a look through and left a few comments inline, but this otherwise LGTM 👍 

Thanks!

c4milo commented:
Thanks for taking the time to review @tombuildsstuff. I've addressed the comments fro the last review. Please let me know if there is anything else you all will want me to correct.

stephybun reviewed (changes requested):
We have a test failure: 
```
------- Stdout: -------
=== RUN   TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition
=== PAUSE TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition
=== CONT  TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition
    testcase.go:113: Step 1/1 error: Error running pre-apply refresh: exit status 1
        Error: Invalid index
          on terraform_plugin_test.tf line 67, in resource "azurerm_kubernetes_cluster" "test":
          67:     vnet_subnet_id = azurerm_subnet.test["19"].id
            ├────────────────
            │ azurerm_subnet.test is tuple with 8 elements
        The given key does not identify an element in this collection value.
        Error: Invalid index
          on terraform_plugin_test.tf line 68, in resource "azurerm_kubernetes_cluster" "test":
          68:     pod_subnet_id  = azurerm_subnet.test["18"].id
            ├────────────────
            │ azurerm_subnet.test is tuple with 8 elements
        The given key does not identify an element in this collection value.
--- FAIL: TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition (12.73s)
FAIL
```

c4milo commented:
@stephybun, I fixed the tests. Thank you! 

```console
❯  make acctests SERVICE='containers' TESTARGS='-run=TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/containers -run=TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition
=== PAUSE TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition
=== CONT  TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition
--- PASS: TestAccKubernetesClusterNodePool_virtualNetworkOwnershipRaceCondition (1163.83s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/containers	1166.633s
```

stephybun reviewed (approved):
Thanks @c4milo LGTM 🍍
